### PR TITLE
mam: last packet - read/write and clean send_ctx map

### PR DESCRIPTION
**Pull request by tsvisabo**

### CODE REVIEW

File: mam/examples/send-msg.c [hunk 1]
```diff
@@ -18,8 +18,14 @@ int main(int ac, char **av) {
   mam_channel_t *channel = NULL;
   retcode_t ret = RC_OK;
 
-  if (ac != 5) {
-    fprintf(stderr, "usage: send-msg <host> <port> <seed> <payload>\n");
+  if (ac != 6) {
+    fprintf(stderr,
+            "usage: send-msg <host> <port> <seed> <payload> <last_packet>\n");
```
Comment (thibault-martinez) [inline mam/examples/send-msg.c:23]: Can you also update the README ?

File: mam/examples/send-packet.c [hunk 2]
```diff
@@ -18,9 +18,15 @@ int main(int ac, char **av) {
   mam_channel_t *channel = NULL;
   retcode_t ret = RC_OK;
 
-  if (ac != 6) {
+  if (ac != 7) {
     fprintf(stderr,
-            "usage: send-packet <host> <port> <seed> <msg_id> <payload>\n");
+            "usage: send-packet <host> <port> <seed> <msg_id> <payload> "
+            "<last_packet>\n");
```
Comment (thibault-martinez) [inline mam/examples/send-packet.c:24]: Can you also update the README ?

File: mam/api/tests/test_api.c [hunk 3]
```diff
@@ -380,6 +381,8 @@ static void test_api_generic() {
     TEST_ASSERT(mam_api_add_psk(&api, pskb) == RC_OK);
   }
 
+  bool is_last_packet;
```
Comment (thibault-martinez) [inline mam/api/tests/test_api.c:384]: Can you initialize to `false` to avoid potential random false-positive ?

File: mam/api/tests/test_api.c [hunk 4]
```diff
@@ -430,27 +435,38 @@ static void test_api_multiple_packets() {
   size_t payload_in_size = strlen((char *)payload_in);
   tryte_t *payload_out = NULL;
   size_t payload_out_size = 0;
+  bool is_last_packet;
```
Comment (thibault-martinez) [inline mam/api/tests/test_api.c:438]: Ditto

File: mam/api/api.c [hunk 5]
```diff
@@ -262,45 +263,68 @@ retcode_t mam_api_bundle_write_packet(
...
+    }
     mam_api_bundle_wrap(bundle, mam_channel_id(ch).p, tag, packet);
+
     trits_free(packet);
   }
 
-  // TODO if last remove pending state
+  if (is_last_packet) {
+    if (!trit_t_to_mam_msg_send_context_t_map_remove(&api->send_ctxs, msg_id)) {
+      return RC_MAM_RECV_CTX_NOT_FOUND;
```
Comment (thibault-martinez) [inline mam/api/api.c]: RECV_CTX ?

File: mam/api/api.c [hunk 6]
```diff
@@ -262,45 +263,68 @@ retcode_t mam_api_bundle_write_packet(
       return RC_OOM;
     }
 
+    if (is_last_packet) {
+      ctx->ord = -ctx->ord;
```
Comment (thibault-martinez) [inline mam/api/api.c:266]: I think negative is only on the tag and not on the actual format ?

File: mam/api/api.c [hunk 7]
```diff
@@ -334,10 +359,23 @@ retcode_t mam_api_bundle_read(mam_api_t *const api,
...
+      curr_tx = (iota_transaction_t *)utarray_eltptr(bundle, packet_index);
+
+      flex_trits_to_trits(tag, NUM_TRITS_TAG, transaction_tag(curr_tx),
+                          NUM_TRITS_TAG, NUM_TRITS_TAG);
+
+      ord =
+          trits_get18(trits_from_rep(MAM_MSG_ORD_SIZE, tag + MAM_MSG_ID_SIZE));
+
+      if (ord < 0) {
+        ctx.ord = ord;
```
Comment (thibault-martinez) [inline mam/api/api.c:370]: Ditto ?

File: mam/api/api.c [hunk 8]
```diff
@@ -356,9 +394,10 @@ retcode_t mam_api_bundle_read(mam_api_t *const api,
     }
 
     ctx = &entry->value;
+    ctx->ord = ord;
```
Comment (thibault-martinez) [inline mam/api/api.c]: It should already have the right value ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/api/tests/test_api.c (hunk 3 above)
Comment (tsvisabo) [reply mam/api/tests/test_api.c:384]: no need, it's been initialized in `mam_api_bundle_read`

File: mam/api/tests/test_api.c (hunk 4 above)
Comment (tsvisabo) [reply mam/api/tests/test_api.c:438]: same

File: mam/api/api.c (hunk 6 above)
Comment (tsvisabo) [reply mam/api/api.c:266]: But then someone can fake a "last packet" bundle?

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**